cli/init.ml

```diff
@@ -0,0 +1,109 @@
+open Duniverse_lib
+open Duniverse_lib.Types
+
+let compute_opam ~repo ~branch ~explicit_root_packages ~excludes ~pins ~remotes =
+  let open Rresult.R.Infix in
+  Opam_cmd.find_local_opam_packages repo >>= fun local_packages ->
+  Opam_cmd.choose_root_packages ~explicit_root_packages ~local_packages >>= fun root_packages ->
+  let root_packages =
+    List.map Opam_cmd.split_opam_name_and_version root_packages |> Opam.sort_uniq
+  in
+  let excludes =
+    List.map Opam_cmd.split_opam_name_and_version (local_packages @ excludes) |> Opam.sort_uniq
+  in
+  Bos.OS.Dir.tmp ".duniverse-opam-root-%s" >>= fun root ->
+  Opam_cmd.init_opam ~root ~remotes () >>= fun () ->
+  Exec.(iter (add_opam_dev_pin ~root) pins) >>= fun () ->
+  Exec.(iter (add_opam_local_pin ~root) local_packages) >>= fun () ->
+  Opam_cmd.calculate_opam ~root ~root_packages ~pins ~excludes ~remotes ~branch >>= fun opam ->
+  let packages_stats = Opam_cmd.packages_stats opam.packages in
+  Opam_cmd.report_packages_stats packages_stats;
+  Ok opam
+
+let compute_repos ~packages =
+  let open Rresult.R.Infix in
+  let packages = Dune_cmd.filter_invalid_packages packages in
+  let dune_packages = List.filter (fun o -> o.Opam.is_dune) packages in
+  Exec.map Dune_cmd.dune_repo_of_opam dune_packages >>= fun repos ->
+  Dune_cmd.dedup_git_remotes repos
+
+let run repo branch explicit_root_packages excludes pins remotes () =
+  let open Rresult.R.Infix in
+  Common.Logs.app "Calculating Duniverse on the %a branch" Styled_pp.branch branch;
+  compute_opam ~repo ~branch ~explicit_root_packages ~excludes ~pins ~remotes >>= fun opam ->
+  let { Opam.root_packages; excludes; pins; remotes; branch; packages } = opam in
+  Common.Logs.app "Calculating Git repositories to vendor";
+  compute_repos ~packages >>= fun repos ->
+  let duniverse = { Duniverse.root_packages; excludes; pins; remotes; branch; packages; repos } in
+  let file = Fpath.(repo // Config.duniverse_file) in
+  Duniverse.save ~file duniverse >>= fun () ->
+  Common.Logs.app "Wrote duniverse file with %a entries to %a."
```

Comment:
Colors are not working, maybe because messages are converted to string by `Common.Logs.app`.

Reply:
Uh good catch!